File: backend/app/schemas/test_quality.py
"""
Test Quality Evaluation Schemas.

Phase 1: 테스트 품질 평가 시스템을 위한 Pydantic 스키마 정의.
"""

from datetime import datetime
from enum import Enum
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ============================================================================
# Enum 정의 (DB CHECK constraint와 일치해야 함)
# ============================================================================


class AnalysisScope(str, Enum):
    """분석 범위."""

    SUBMISSION = "submission"
    PROBLEM_RUBRIC = "problem_rubric"


class AnalysisStatus(str, Enum):
    """분석 상태."""

    PENDING = "PENDING"
    RUNNING = "RUNNING"
    SUCCESS = "SUCCESS"
    ERROR = "ERROR"


class QualityGrade(str, Enum):
    """품질 등급."""

    A = "A"
    B = "B"
    C = "C"
    D = "D"
    F = "F"


# ============================================================================
# AntiPattern 관련 스키마
# ============================================================================


class AntiPatternDetail(BaseModel):
    """감지된 안티패턴 상세."""

    type: str = Field(..., description="안티패턴 유형")
    penalty: int = Field(..., description="감점")
    location: str = Field(..., description="테스트 함수명")
    line: Optional[int] = Field(None, description="라인 번호")
    description: Optional[str] = Field(None, description="설명")


# ============================================================================
# Per-Test 분류 결과
# ============================================================================


class TestCaseAnalysis(BaseModel):
    """단일 테스트 케이스 분석 결과."""

    name: str = Field(..., description="테스트 함수명")
    value_types: list[str] = Field(default_factory=list, description="값 유형 목록")
    input_diversities: list[str] = Field(
        default_factory=list, description="입력 다양성 목록"
    )
    test_purposes: list[str] = Field(default_factory=list, description="테스트 목적 목록")
    antipatterns: list[AntiPatternDetail] = Field(
        default_factory=list, description="안티패턴 목록"
    )
    confidence: float = Field(1.0, ge=0.0, le=1.0, description="분류 신뢰도")


# ============================================================================
# 점수 세부 내역
# ============================================================================


class ScoringBreakdown(BaseModel):
    """점수 계산 세부 내역."""

    value_type_score: float = Field(0.0, description="값 유형 점수")
    input_diversity_score: float = Field(0.0, description="입력 다양성 점수")
    test_purpose_score: float = Field(0.0, description="테스트 목적 점수")
    antipattern_penalty: float = Field(0.0, description="안티패턴 감점")
    final_score: float = Field(..., description="최종 점수")


# ============================================================================
# 제출 분석 결과 (test_quality_analysis JSONB)
# ============================================================================


class TestQualityAnalysis(BaseModel):
    """테스트 품질 분석 결과 (JSONB 저장용)."""

    parser_version: str = Field(..., description="파서 버전")
    scoring_version: str = Field(..., description="점수 모델 버전")
    test_count: int = Field(..., ge=0, description="테스트 함수 수")
    effective_test_count: int = Field(
        ..., ge=0, description="유효 테스트 케이스 수 (parametrize 확장)"
    )
    value_types: list[str] = Field(default_factory=list, description="전체 값 유형")
    input_diversities: list[str] = Field(
        default_factory=list, description="전체 입력 다양성"
    )
    test_purposes: list[str] = Field(default_factory=list, description="전체 테스트 목적")
    antipatterns: list[AntiPatternDetail] = Field(
        default_factory=list, description="감지된 안티패턴"
    )
    per_test: list[TestCaseAnalysis] = Field(
        default_factory=list, description="테스트별 상세"
    )
    scoring_breakdown: Optional[ScoringBreakdown] = Field(None, description="점수 세부")
    overall_confidence: float = Field(1.0, ge=0.0, le=1.0, description="전체 신뢰도")

    model_config = {"from_attributes": True}


# ============================================================================
# 문제 루브릭 분석 결과 (rubric_analysis JSONB)
# ============================================================================


class ExpectedCategories(BaseModel):
    """기대되는 카테고리."""

    value_types: list[str] = Field(default_factory=list)
    input_diversities: list[str] = Field(default_factory=list)
    test_purposes: list[str] = Field(default_factory=list)


class RubricWeights(BaseModel):
    """루브릭 가중치."""

    value_type_coverage: float = Field(0.35, ge=0.0, le=1.0)
    diversity_coverage: float = Field(0.30, ge=0.0, le=1.0)
    purpose_coverage: float = Field(0.35, ge=0.0, le=1.0)


class RubricAnalysis(BaseModel):
    """문제 루브릭 분석 결과 (JSONB 저장용)."""

    parser_version: str = Field(...)
    scoring_version: str = Field(...)
    expected_categories: ExpectedCategories = Field(...)
    weights: RubricWeights = Field(default_factory=RubricWeights)
    difficulty_factor: float = Field(1.0, ge=0.5, le=2.0)

    model_config = {"from_attributes": True}


# ============================================================================
# AnalysisRun 스키마
# ============================================================================


class AnalysisRunBase(BaseModel):
    """분석 실행 기본 스키마."""

    scope: AnalysisScope
    parser_version: str = Field(..., max_length=10)
    scoring_version: str = Field(..., max_length=10)
    source_hash: str = Field(..., max_length=32, description="소스 코드 MD5 해시")


class AnalysisRunCreate(AnalysisRunBase):
    """분석 실행 생성 스키마."""

    submission_id: Optional[UUID] = None
    problem_id: Optional[int] = None


class AnalysisRunResponse(AnalysisRunBase):
    """분석 실행 응답 스키마."""

    id: int
    submission_id: Optional[UUID] = None
    problem_id: Optional[int] = None
    status: AnalysisStatus
    confidence_score: Optional[float] = None
    result: Optional[dict[str, Any]] = None
    error_message: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ============================================================================
# 제출 품질 응답 확장
# ============================================================================


class SubmissionQualityResponse(BaseModel):
    """제출 품질 평가 응답."""

    submission_id: UUID
    test_quality_score: Optional[float] = None
    test_quality_grade: Optional[QualityGrade] = None
    test_quality_analysis: Optional[TestQualityAnalysis] = None
    analysis_run: Optional[AnalysisRunResponse] = None

    model_config = {"from_attributes": True}


# ============================================================================
# 문제 루브릭 응답
# ============================================================================


class ProblemRubricResponse(BaseModel):
    """문제 루브릭 평가 응답."""

    problem_id: int
    rubric_score: Optional[float] = None
    rubric_analysis: Optional[RubricAnalysis] = None
    analysis_run: Optional[AnalysisRunResponse] = None

    model_config = {"from_attributes": True}


# ============================================================================
# 통계 응답
# ============================================================================


class QualityStatisticsResponse(BaseModel):
    """품질 통계 응답."""

    grade_distribution: dict[str, int] = Field(
        default_factory=dict, description="등급별 분포"
    )
    average_score: float = Field(0.0, description="평균 점수")
    analyzed_submissions: int = Field(0, description="분석 완료 제출 수")

    model_config = {"from_attributes": True}

File: frontend/types/test-quality.ts
/**
 * Test Quality Evaluation Types.
 *
 * Phase 4-1: 테스트 품질 평가 시스템을 위한 TypeScript 타입 정의.
 * Backend schemas/test_quality.py와 동기화.
 */

// ============================================================================
// Enum 정의
// ============================================================================

export type AnalysisScope = "submission" | "problem_rubric";

export type AnalysisStatus = "PENDING" | "RUNNING" | "SUCCESS" | "ERROR";

export type QualityGrade = "A" | "B" | "C" | "D" | "F";

// ============================================================================
// 카테고리 정의
// ============================================================================

export type ValueType =
  | "normal"
  | "boundary"
  | "edge"
  | "negative"
  | "extreme"
  | "unknown";

export type InputDiversity =
  | "single"
  | "multiple"
  | "empty"
  | "duplicate"
  | "mixed_types"
  | "unknown";

export type TestPurpose =
  | "happy_path"
  | "error_handling"
  | "boundary_check"
  | "equivalence_partition"
  | "state_transition"
  | "unknown";

// ============================================================================
// 카테고리 한글 매핑
// ============================================================================

export const VALUE_TYPE_LABELS: Record<ValueType, string> = {
  normal: "정상값",
  boundary: "경계값",
  edge: "엣지케이스",
  negative: "음수/비정상",
  extreme: "극단값",
  unknown: "미분류",
};

export const INPUT_DIVERSITY_LABELS: Record<InputDiversity, string> = {
  single: "단일 입력",
  multiple: "복수 입력",
  empty: "빈 입력",
  duplicate: "중복 요소",
  mixed_types: "혼합 타입",
  unknown: "미분류",
};

export const TEST_PURPOSE_LABELS: Record<TestPurpose, string> = {
  happy_path: "정상 경로",
  error_handling: "에러 처리",
  boundary_check: "경계 검사",
  equivalence_partition: "동치 분할",
  state_transition: "상태 전이",
  unknown: "미분류",
};

export const GRADE_LABELS: Record<QualityGrade, string> = {
  A: "우수",
  B: "양호",
  C: "보통",
  D: "미흡",
  F: "부족",
};

export const GRADE_COLORS: Record<QualityGrade, string> = {
  A: "text-green-600 dark:text-green-400",
  B: "text-blue-600 dark:text-blue-400",
  C: "text-yellow-600 dark:text-yellow-400",
  D: "text-orange-600 dark:text-orange-400",
  F: "text-red-600 dark:text-red-400",
};

export const GRADE_BG_COLORS: Record<QualityGrade, string> = {
  A: "bg-green-100 dark:bg-green-900/30",
  B: "bg-blue-100 dark:bg-blue-900/30",
  C: "bg-yellow-100 dark:bg-yellow-900/30",
  D: "bg-orange-100 dark:bg-orange-900/30",
  F: "bg-red-100 dark:bg-red-900/30",
};

// ============================================================================
// AntiPattern 관련 타입
// ============================================================================

export interface AntiPatternDetail {
  type: string;
  penalty: number;
  location: string;
  line?: number;
  description?: string;
}

export const ANTIPATTERN_LABELS: Record<string, string> = {
  NO_ASSERTION: "Assertion 없음",
  EXCEPTION_SWALLOWED: "예외 무시",
  MAGIC_NUMBER: "매직 넘버",
  DUPLICATE_TEST: "중복 테스트",
  EMPTY_TEST: "빈 테스트",
};

// ============================================================================
// Per-Test 분류 결과
// ============================================================================

export interface TestCaseAnalysis {
  name: string;
  value_types: string[];
  input_diversities: string[];
  test_purposes: string[];
  antipatterns: AntiPatternDetail[];
  confidence: number;
}

// ============================================================================
// 점수 세부 내역
// ============================================================================

export interface ScoringBreakdown {
  value_type_score: number;
  input_diversity_score: number;
  test_purpose_score: number;
  antipattern_penalty: number;
  final_score: number;
}

// ============================================================================
// 제출 분석 결과
// ============================================================================

export interface TestQualityAnalysis {
  parser_version: string;
  scoring_version: string;
  test_count: number;
  effective_test_count: number;
  value_types: string[];
  input_diversities: string[];
  test_purposes: string[];
  antipatterns: AntiPatternDetail[];
  per_test: TestCaseAnalysis[];
  scoring_breakdown?: ScoringBreakdown;
  overall_confidence: number;
}

// ============================================================================
// 문제 루브릭 분석 결과
// ============================================================================

export interface ExpectedCategories {
  value_types: string[];
  input_diversities: string[];
  test_purposes: string[];
}

export interface RubricWeights {
  value_type_coverage: number;
  diversity_coverage: number;
  purpose_coverage: number;
}

export interface RubricAnalysis {
  parser_version: string;
  scoring_version: string;
  expected_categories: ExpectedCategories;
  weights: RubricWeights;
  difficulty_factor: number;
}

// ============================================================================
// AnalysisRun 타입
// ============================================================================

export interface AnalysisRunResponse {
  id: number;
  submission_id?: string;
  problem_id?: number;
  scope: AnalysisScope;
  status: AnalysisStatus;
  parser_version: string;
  scoring_version: string;
  source_hash: string;
  confidence_score?: number;
  result?: Record<string, unknown>;
  error_message?: string;
  created_at: string;
}

// ============================================================================
// API 응답 타입
// ============================================================================

export interface SubmissionQualityResponse {
  submission_id: string;
  test_quality_score?: number;
  test_quality_grade?: QualityGrade;
  test_quality_analysis?: TestQualityAnalysis;
  analysis_run?: AnalysisRunResponse;
}

export interface ProblemRubricResponse {
  problem_id: number;
  rubric_score?: number;
  rubric_analysis?: RubricAnalysis;
  analysis_run?: AnalysisRunResponse;
}

export interface QualityStatisticsResponse {
  grade_distribution: Record<string, number>;
  average_score: number;
  analyzed_submissions: number;
}

// ============================================================================
// 힌트 관련 타입 (Phase 3-2)
// ============================================================================

export type HintDifficulty = "easy" | "medium" | "hard";

export interface TestHint {
  category: string;
  hint_text: string;
  difficulty: HintDifficulty;
  question: string;
}

export interface HintGenerationResult {
  summary: string;
  hints: TestHint[];
  encouragement: string;
  next_step: string;
}

// ============================================================================
// 테스트 생성 관련 타입 (Phase 3-1, Admin용)
// ============================================================================

export interface GeneratedTest {
  test_name: string;
  test_code: string;
  target_categories: string[];
  rationale: string;
}

export interface CoverageImprovement {
  value_types_added: string[];
  diversities_added: string[];
  purposes_added: string[];
}

export interface TestGenerationResult {
  generated_tests: GeneratedTest[];
  complete_test_file: string;
  coverage_improvement: CoverageImprovement;
}

// ============================================================================
// 유틸리티 함수
// ============================================================================

/**
 * 점수에 따른 등급 반환
 */
export function getGradeFromScore(score: number): QualityGrade {
  if (score >= 90) return "A";
  if (score >= 75) return "B";
  if (score >= 60) return "C";
  if (score >= 40) return "D";
  return "F";
}

/**
 * 등급에 따른 색상 클래스 반환
 */
export function getGradeColor(grade: QualityGrade): string {
  return GRADE_COLORS[grade];
}

/**
 * 등급에 따른 배경색 클래스 반환
 */
export function getGradeBgColor(grade: QualityGrade): string {
  return GRADE_BG_COLORS[grade];
}

/**
 * 점수에 따른 색상 반환
 */
export function getScoreColor(score: number): string {
  if (score >= 90) return "text-green-600 dark:text-green-400";
  if (score >= 70) return "text-blue-600 dark:text-blue-400";
  if (score >= 50) return "text-yellow-600 dark:text-yellow-400";
  return "text-red-600 dark:text-red-400";
}

/**
 * 카테고리 커버리지 비율 계산
 */
export function calculateCoverage(
  covered: string[],
  total: number
): { count: number; percentage: number } {
  const uniqueCovered = new Set(
    covered.filter((c) => c.toLowerCase() !== "unknown")
  );
  return {
    count: uniqueCovered.size,
    percentage: total > 0 ? (uniqueCovered.size / total) * 100 : 0,
  };
}

// 전체 카테고리 수
export const TOTAL_VALUE_TYPES = 5;
export const TOTAL_INPUT_DIVERSITIES = 5;
export const TOTAL_TEST_PURPOSES = 5;
